### PR TITLE
Ecm 94 Session Days Report

### PR DESCRIPTION
**Pull request by gauravchawla73463**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/ECM-94

### Change description ###
Session Days report implemented

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+    private List<HearingTypeItem> getHearings(SessionDaysCaseData caseData) {
+        var hearings = caseData.getHearingCollection();
+        if (hearings == null) {
+            return Collections.emptyList();
+        }
+        return hearings;
+    }
+
+    private void setReportDetail(SessionDaysCaseData caseData, List<SessionDaysReportDetail> reportDetailList) {
+        for (HearingTypeItem h : getHearings(caseData)) {
```
Comment (hsjhita) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java:181]: personally, would change the name of 'h' into something more meaningful, maybe refactor to hearingTypeItem? same with 'd' below

**Review by hsjhita — APPROVED** [review]

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java (hunk 1 above)
Comment (gauravchawla73463) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java]: refactored

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+    private void executeReport(List<SessionDaysSubmitEvent> submitEvents, SessionDaysReportData sessionDaysReportData) {
+        setReportData(submitEvents, sessionDaysReportData);
+    }
+
+    private void setReportData(List<SessionDaysSubmitEvent> submitEvents, SessionDaysReportData reportData) {
+        List<SessionDaysReportSummary2> sessionDaysReportSummary2List = new ArrayList<>();
+        List<SessionDaysReportDetail> sessionDaysReportDetailList = new ArrayList<>();
+        for (SessionDaysSubmitEvent submitEvent : submitEvents) {
+            var c = submitEvent.getCaseData();
```
[the remark this answers is not in the document]
Comment (gauravchawla73463) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java]: refactored

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+        for (HearingTypeItem hearingTypeItem : getHearings(caseData)) {
+            var dates = hearingTypeItem.getValue().getHearingDateCollection();
+            if (CollectionUtils.isNotEmpty(dates)) {
+                for (DateListedTypeItem dateListedTypeItem : dates) {
+                    if (isHearingStatusValid(dateListedTypeItem)) {
+                        String judgeStatus = getJudgeStatus(hearingTypeItem.getValue().getJudge());
+                        SessionDaysReportSummary2 reportSummary2 = getReportSummary2Item(
+                                dateListedTypeItem.getValue(), sessionDaysReportSummary2List);
+                        switch (judgeStatus) {
+                            case "SALARIED":
```
Comment (jthmcts) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java:131]: Having created an enum for judge employment status then you really ought to use it

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+                }
+            }
+        }
+    }
+
+    private boolean isHearingStatusValid(DateListedTypeItem dateListedTypeItem) {
+        return HEARING_STATUS_HEARD.equals(dateListedTypeItem.getValue().getHearingStatus());
+    }
+
+    private String getJudgeStatus(String judgeName) {
```
Comment (jthmcts) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java:164]: Why not use the enum

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+            var hearingEndTime = LocalDateTime.parse(hearingTimingFinish, OLD_DATE_TIME_PATTERN);
+            long startToEndDiffInMinutes = ChronoUnit.MINUTES.between(hearingStartTime, hearingEndTime);
+            duration = startToEndDiffInMinutes - breakDuration;
+        }
+
+        return String.valueOf(duration);
+    }
+
+    private String getSessionType(String hearingDuration) {
+        String sessionType = NONE;
```
Comment (jthmcts) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java:259]: You don't need this really. Just return the value directly

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+        var duration = Long.parseLong(hearingDuration);
+        if (duration > 0 && duration < 60) {
+            sessionType = ONE_HOUR;
+        } else if (duration >= 60 && duration <= 180) {
+            sessionType = HALF_DAY;
+        } else if (duration > 180) {
+            sessionType = FULL_DAY;
+        }
+        return sessionType;
```
Comment (jthmcts) [inline src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java:268]:
```
else {
   return NONE;
}
```

File: src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/ReportDocHelperTest.java [hunk 7]
```diff
@@ -1,13 +1,13 @@
 package uk.gov.hmcts.ethos.replacement.docmosis.helpers;
 
 import com.fasterxml.jackson.databind.ObjectMapper;
-import java.util.Arrays;
-import java.util.Collections;
+import java.util.*;
```
Comment (jthmcts) [inline src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/ReportDocHelperTest.java:4]: checkstyle

File: src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReportTest.java [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+        return judge;
+    }
+
+    @Test
+    public void shouldNotShowCaseWithNoHearings() {
+        // Given a case has no hearing
+        // and report data is requested
+        // the case should not be in the report data
+
+            caseDataBuilder.withNoHearings();
```
Comment (jthmcts) [inline src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReportTest.java:65]: indentation looks wonky

File: src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReportTest.java [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+        assertEquals("111", reportDetail.getCaseReference());
+        assertEquals("Clerk A", reportDetail.getHearingClerk());
+        assertEquals("2022-01-20", reportDetail.getHearingDate());
+        assertEquals("1", reportDetail.getHearingNumber());
+        assertEquals("Y", reportDetail.getHearingSitAlone());
+        assertEquals("Y", reportDetail.getHearingTelConf());
+        assertEquals(FULL_DAY, reportDetail.getSessionType());
+        assertEquals(judge, reportDetail.getHearingJudge());
+        assertEquals(judgeType, reportDetail.getJudgeType());
+
```
Comment (jthmcts) [inline src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReportTest.java:148]: +
+

**Review by jthmcts — CHANGES_REQUESTED** [review]
This needs a SQL script to create the table

File: src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/ReportDocHelperTest.java (hunk 7 above)
Comment (gauravchawla73463) [reply src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/helpers/ReportDocHelperTest.java:4]: fixed

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java (hunk 3 above)
Comment (gauravchawla73463) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java]: using enum now

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java (hunk 4 above)
Comment (gauravchawla73463) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java]: refactored

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java (hunk 5 above)
Comment (gauravchawla73463) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java]: refactored

File: src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java (hunk 6 above)
Comment (gauravchawla73463) [reply src/main/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReport.java]: refactored

File: src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReportTest.java (hunk 8 above)
Comment (gauravchawla73463) [reply src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReportTest.java:68]: fixed

File: src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReportTest.java (hunk 9 above)
Comment (gauravchawla73463) [reply src/test/java/uk/gov/hmcts/ethos/replacement/docmosis/reports/sessiondays/SessionDaysReportTest.java:151]: fixed

**Review by jthmcts — APPROVED** [review]

**Pull request merged**